docs/guides/basic-rbac.mdx

```diff
@@ -3,267 +3,303 @@ title: Implement basic Role Based Access Control (RBAC) with metadata
 
-  return (
-    <>
-      <h1>This is the admin dashboard</h1>
-      <p>This page is restricted to users with the 'admin' role.</p>
-    </>
-  )
-}
-```
+  ```ts {{ filename: 'app/admin/_actions.ts' }}
+  'use server'
 
-> [!NOTE]
-> You can modify the behavior of the helper function to meet your needs. Maybe it will return the roles that the user has, or you could create a `protectByRole()` and have that function handle the redirect.
+  import { checkRole } from '@/utils/roles'
+  import { clerkClient } from '@clerk/nextjs/server'
 
-## Add admin tools to find users and add roles
+  export async function setRole(formData: FormData) {
+    // Check that the user trying to set the role is an admin
+    if (!checkRole('admin')) {
+      return { message: 'Not Authorized' }
+    }
 
-You can leverage the `checkRole()` function you added along with server actions to build basic tools for finding users and managing roles.
+    try {
+      const res = await clerkClient().users.updateUser(formData.get('id') as string, {
+        publicMetadata: { role: formData.get('role') },
+      })
+      return { message: res.publicMetadata }
+    } catch (err) {
+      return { message: err }
+    }
+  }
+  ```
+
+  With the server action set up, now build the `<SearchUsers />` component. This component includes a form for searching users, and when submitted, appends the search term to the URL as a search parameter. Your `/admin` route will then perform a query based on the updated URL.
+
+  1. In your `app/admin/` directory, create a `_search-users.tsx` file.
+  1. Add the following code to the file.
+
+  ```tsx {{ filename: 'app/admin/_search-users.tsx' }}
+  'use client'
+
+  import { usePathname, useRouter } from 'next/navigation'
+
+  export const SearchUsers = () => {
+    const router = useRouter()
+    const pathname = usePathname()
+
+    return (
+      <div>
+        <form
+          onSubmit={(e) => {
+            e.preventDefault()
+            const form = e.currentTarget
+            const formData = new FormData(form)
+            const queryTerm = formData.get('search') as string
+            router.push(pathname + '?search=' + queryTerm)
+          }}
+        >
+          <label htmlFor="search">Search for Users</label>
+          <input id="search" name="search" type="text" />
+          <button type="submit">Submit</button>
+        </form>
+      </div>
+    )
+  }
+  ```
 
-Start with the server action. The `setRole()` action below will check the role of the current user using the `checkRoles()` helper to confirm that the user is an `admin`. It will then set the role for the selected user to the specified role.
+  With the server action and the search form set up, it's time to refactor the `app/admin/page.tsx` to be a server component. It will check whether a search parameter has been appended to the URL by the search form. If a search parameter is present, it will query for users matching the entered term.
```

Comment:
```suggestion
  With the server action and the search form set up, it's time to refactor the `app/admin/page.tsx`. It will check whether a search parameter has been appended to the URL by the search form. If a search parameter is present, it will query for users matching the entered term.
```

The admin page was already a server component.